### PR TITLE
GROOVY-8977: Collect default imported class info for better resolving…

### PR DESCRIPTION
**Pull request by danielsun1106**
… performance

### CODE REVIEW

File: src/main/java/org/codehaus/groovy/control/DefaultImportedClassCollectorHelper.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        private static final File CLASSES_INFO_FILE = new File(new File(GroovySystem.getGroovyRoot(), "caches"), "default_imported_classes");
+        private static final String JAVA_VERSION = System.getProperty("java.version");
+        private static final String GROOVY_VERSION = getGroovyVersion();
+
+        private static String getGroovyVersion() {
+            String version = GroovySystem.getVersion();
+            version = (null == version || version.trim().isEmpty()) ? SNAPSHOT : version;
+
+            // if using SNAPSHOT version(e.g. 3.0.0-SNAPSHOT), cache for only 1 hour because default imported classes may change when developing
+            return version.contains(SNAPSHOT) ? version + new SimpleDateFormat("-yyyyMMdd.HH").format(new Date()) : version;
```
Comment (melix) [inline src/main/java/org/codehaus/groovy/control/DefaultImportedClassCollectorHelper.java]: :x: I think this is problematic. We shouldn't bake in a rule like this. Ideally this should be a hash of the Groovy runtime.

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/control/DefaultImportedClassCollectorHelper.java]: "Ideally this should be a hash of the Groovy runtime." -- could you provide me more details about Groovy runtime?

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/control/DefaultImportedClassCollectorHelper.java]: @melix  "I would also avoid adding another third party dependency just for this: this is core groovy." -- Yep, I agree with you. so the third party dependency is optional, as you can see, the dependency is `compileOnly`

**Comment by danielsun1106** [comment]
@melix "I would also avoid adding another third party dependency just for this: this is core groovy." -- Yep, I agree with you. so the third party dependency is optional, as you can see, the dependency is `compileOnly` and it will not be included in groovy distributions.
```groovy
compileOnly "io.github.classgraph:classgraph:$classgraphVersion" // for compilation only
testRuntime "io.github.classgraph:classgraph:$classgraphVersion" // for running tests
```

**Pull request closed without merging**